### DISCUSSION TITLE
better custom widgets with nested widgets

### DISCUSSION THREAD

**Discussion by JumpyLionnn** [emilk/egui#3195] [Q&A]
I'd like to create a custom widget consisting of a rectangle with a text edit on top of it. Additionally, I want to enable click interactions with this rectangle. The rectangle is bigger than the text edit.
Please note that the example I'm providing here is simplified.
I know i can use the `ui.allocate_exact_size` function to get an interactable rect. If I do that, I can't add ui widgets on top of it.
if I calculate the rect I need before, i allocate the space and then I use `ui.allocate_ui_at_rect` then, the interactable rect I get from it is the min size of the inner widgets and not the size of the rect I passed in.
Currently, I allocate size, I move the pointer up in an hacky way (with negative value for `ui.add_space` with the sum of the rect height and the item spacing) and then I draw the ui. I really don't like this way. I don't think the `add_space` function is meant for that. I also don't think it should have negative values and I don't think its suitable for that kind of task.

is there another way i can do that? if not ill just open an issue for that

**Comment by jngls** [comment]
I also need to be able to nest widgets inside each other. For example, I have a custom checkbox widget, and then another widget containing a cluster of checkboxes which gets reused elsewhere. Finding this to be extremely hard.

**Comment by YgorSouza** [comment]
If you just want to place widgets inside a previously allocated rect, I think the method for that is [Ui::put](https://docs.rs/egui/0.24.1/egui/struct.Ui.html#method.put). Then the `Response` from the background rect will only be interacted with when the pointer is on it, and not on the widgets, but if you want to consider the response from the inner widgets as well, you can make a union of them using the `BitOr` operator.

```rs
impl Default for MyApp {
    fn default() -> Self {
        Self {
            name: "Arthur".to_owned(),
        }
    }
}

impl eframe::App for MyApp {
    fn update(&mut self, ctx: &egui::Context, _frame: &mut eframe::Frame) {
        egui::CentralPanel::default().show(ctx, |ui| {
            let (rect, response) = ui.allocate_exact_size([300.0; 2].into(), egui::Sense::click());
            if response.hovered() {
                ui.painter()
                    .rect(rect, 0.0, egui::Color32::BLUE, egui::Stroke::NONE);
            }
            let inner_response = ui.put(rect, |ui: &mut egui::Ui| {
                ui.vertical(|ui| {
                    (0..10)
                        .map(|_| ui.text_edit_singleline(&mut self.name))
                        .reduce(std::ops::BitOr::bitor)
                        .unwrap()
                })
                .inner
            });
            if inner_response.hovered() {
                ui.painter().rect(
                    rect,
                    0.0,
                    egui::Color32::TRANSPARENT,
                    egui::Stroke::new(2.0, egui::Color32::GREEN),
                );
            }
        });
    }
}
```
[Screencast from 2023-12-20 20-18-38.webm](https://github.com/emilk/egui/assets/43298013/5933ccb3-726e-4db3-9322-81a58d7162a8)


However, I noticed that `put()` changes the position of the cursor, so the next widget isn't necessarily drawn after the rect you allocated, which I think is a bug, unless I'm misunderstanding how it is supposed to work. I'll open an issue about it and see if anyone says anything.

Edit: `put()` just calls [centered_and_justified](https://docs.rs/egui/0.24.1/egui/struct.Ui.html#method.centered_and_justified), which specifies that you should only give it one widget, and not a vertical layout like I did, so I guess it is not a bug. But maybe it could be changed to suit this use case.
